u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java

```diff
@@ -192,6 +194,85 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
     }
   }
 
+  public static TransferAccessResponse decodeTransferAccessResponse(byte[] data)
+      throws U2FException {
+    int sequenceNumberSize = 1;
+    int publicKeySize = 65;
+    int appIdSize = 32;
+
+    try {
+      DataInputStream inputStream = new DataInputStream(new ByteArrayInputStream(data));
+      byte controlByte = inputStream.readByte();
+
+      if (!inputStream.markSupported()) {
+        throw new U2FException("Stream doesn't support mark and reset.");
+      }
+      inputStream.mark(inputStream.available());
+      int numberOfTransferAccessMessages = inputStream.readUnsignedByte();
+      inputStream.reset();
+
+      TransferAccessMessage[] transferAccessMessages =
+          new TransferAccessMessage[numberOfTransferAccessMessages];
+
+      for (int i = 0; i < numberOfTransferAccessMessages; i++) {
+        inputStream.mark(inputStream.available());
+        inputStream.skipBytes(sequenceNumberSize + publicKeySize + appIdSize);
+
+        X509Certificate newAttestationCertificate = (X509Certificate) CertificateFactory
+            .getInstance("X.509").generateCertificate(inputStream);
+        byte[] newAttestationCertificateBytes;
+        try {
+          newAttestationCertificateBytes = newAttestationCertificate.getEncoded();
+        } catch (CertificateEncodingException e) {
+          throw new U2FException("Error when encoding attestation certificate.", e);
+        }
+
+        int lengthOfAttestationCertificate = newAttestationCertificateBytes.length;
+
+        int lengthOfSignatureUsingPrivateKey = inputStream.readUnsignedByte();
+        inputStream.skipBytes(lengthOfSignatureUsingPrivateKey);
+        int lengthOfSignatureUsingAttestationKey = inputStream.readUnsignedByte();
+        inputStream.reset();
+
+        int lengthField_signatureUsingAttestationKey = 1;
+        int lengthField_signatureUsingPrivateKey = 1;
+        int lengthOfRawTransferAccessMessage =
+            sequenceNumberSize + publicKeySize + appIdSize + lengthOfAttestationCertificate
+                + lengthField_signatureUsingPrivateKey + lengthOfSignatureUsingPrivateKey
+                + lengthField_signatureUsingAttestationKey + lengthOfSignatureUsingAttestationKey;
+
+        byte[] rawTransferAccessMessage = new byte[lengthOfRawTransferAccessMessage];
+        inputStream.readFully(rawTransferAccessMessage);
+        transferAccessMessages[numberOfTransferAccessMessages - i - 1] =
```

Comment:
things like -1 make me nervous.  add some comments to explain what you're doing.